Complete custom email notification batching period feature.

abhijeetbodas2001 opened the pull request:
Fixes #15280

## Screenshots

![image](https://user-images.githubusercontent.com/55339528/129443725-33cc6a56-8980-4fcf-9283-1076b82f9b69.png)
---
![image](https://user-images.githubusercontent.com/55339528/129443770-701bc3e7-c3d1-40df-ac61-cf6b493c99d0.png)

abhijeetbodas2001 commented:
@sahil839 Could you review this?

sahil839 reviewed (commented):
Left a minor comment otherwise the code LGTM.
We can do some changes to UX though -
- We can disable the setting if email notifications are disabled.
- I have not seen the previous discussion of this, but we can have an option of setting custom period, but this should not be a blocker and can be added later also.

abhijeetbodas2001 commented on static/js/settings.js:
```diff
@@ -55,6 +55,9 @@ function setup_settings_label() {
         }),
+        email_notifications_batching_period: $t({
+            defaultMessage: "Batching period for email notifications",
+        }),
```
Removed this field, thanks!

abhijeetbodas2001 commented:
Thanks for the review! (and sorry for the slow response 😅) I've updated the PR after addressing the comment above. For the UX:

> We can disable the setting if email notifications are disabled.

I tried doing that (after checking out how that's implemented for the "notification sound" dropdown), and found that we have a bug in the similar logic for the "select notification sound" dropdown:
```js
    const notification_sound_dropdown = $("#notification_sound");
    notification_sound_dropdown.val(user_settings.notification_sound);

    $("#enable_sounds, #enable_stream_audible_notifications").on("change", () => {
        if (
            $("#enable_stream_audible_notifications").prop("checked") ||
            $("#enable_sounds").prop("checked")
        ) {
            notification_sound_dropdown.prop("disabled", false);
            notification_sound_dropdown.parent().removeClass("control-label-disabled");
        } else {
            notification_sound_dropdown.prop("disabled", true);
            notification_sound_dropdown.parent().addClass("control-label-disabled");
        }
    });
```

This code ignores the stream-specific rows, and hence is buggy. If someone enables the audible notifications for a stream while the global settings have been disabled, the dropdown remains disabled.

I'm not sure how to fix this, since that would involve, after every change of checkbox state checking all the other check boxes of the stream specific rows. Any suggestions?

---

> I have not seen the previous discussion of this, but we can have an option of setting custom period, but this should not be a blocker and can be added later also.

I don't remember discussing that aspect before, but it looks like Microsoft teams uses a similar UI for what seems like their version of this feature:
![image](https://user-images.githubusercontent.com/55339528/130072992-7b3009ce-becc-4735-ab09-84568141c273.png)

And we could easily add a couple more values to the dropdown to increase the granularity if we wanted to.

sahil839 commented:
Hmm, yes checking stream-specific settings will be a problem. Maybe we can add a class ".email-notifications" to all the email checkboxes (including stream-specific ones) and then check something like `$(".email_notifications:checked").length > 0` or `$(".email_notifications").is("checked")> 0`, not sure about performance aspects of using these though.

And we can wait for some more feedback on adding other/custom option.

timabbott commented:
We should have a custom option; you can look at the "message edit deadline" dropdowns for an implementation template to crib from.

I think we can leave the "disable dropdown" feature for follow-up work (maybe worth a CZO thread); it's not clear it'd be net helpful even if we can fix the live update issues discussed above.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5014,6 +5015,26 @@ def do_create_realm(
+        + datetime.timedelta(seconds=new_batching_period)
+    )
+
+    existing_scheduled_emails.update(scheduled_timestamp=new_scheduled_timestamp)
```
Can we write this using `Q` to do an UPDATE query to add a delta to all matching rows without fully fetching the objects and without assuming that the `scheduled_timestamp` is the same for all of these?

That will make this a bit less fragile around possible future changes to the model.

abhijeetbodas2001 replied:
You mean `F`, right? I updated the commit to use that.

timabbott replied:
Yeah, that's what I meant; I can never keep Q and F straight :)

timabbott commented on zerver/tests/test_settings.py:
```diff
@@ -190,6 +199,54 @@ def test_change_email_batching_period(self) -> None:
+        hamlet = self.example_user("hamlet")
+        self.assertEqual(hamlet.email_notifications_batching_period_seconds, 10 * 60)
+
+        def check_shceduled_timestamp(entry_id: int, expected_timestamp: datetime) -> None:
```
typo in this function name.

abhijeetbodas2001 replied:
Fixed!

timabbott commented on zerver/tests/test_settings.py:
```diff
@@ -190,6 +199,54 @@ def test_change_email_batching_period(self) -> None:
+
+        # Update Hamlet's setting from 300 seconds (5 minutes) to 600 seconds (10 minutes)
+        result = self.client_patch(
+            "/json/settings", {"email_notifications_batching_period_seconds": 10 * 60}
```
Can you add an assertion that the current batching period was 5*60 before this request?  Will make the test more readable.

abhijeetbodas2001 replied:
Done

timabbott commented:
This lgtm but can you also add /help/ updates covering this feature?  We can do a new brief article but @alya FYI we may want to later merge into a large "Email notification configuration" page.

abhijeetbodas2001 commented:
@timabbott Thanks for the review! This isn't ready for another round yet. I've addressed comments on the first commit, and will do the frontend/docs changes as soon as possible.

timabbott commented:
I rebased the first commit and merged it as 17b8d5361225c92625ef951dfc263592f11d2f1d.

timabbott commented:
Since this is getting stale and essentially complete, I went ahead and did enough that I think we can merge the frontend as well.  

@alya want to do a final round of review on the frontend for this (wording, etc.), opening issues for what remains to do (you can tag them as release goals) and doing a documentation PR?

timabbott commented:
Opened https://github.com/zulip/zulip/issues/19713 for one of the follow-ups.

timabbott commented:
Merged, thanks @abhijeetbodas2001!

alya commented:
Awesome!

User doc PR: #19714

abhijeetbodas2001 commented:
@timabbott @alya Thanks for all your work, and apologies for having needed to do the rebasing and documentation! I got busy with assignments and mid-term exams, so could find time.
I'll try to finish off #19713 this coming weekend.

alya commented:
No worries, we really appreciate your work! Hope the exams went well! :)
